examples/pytorch/zero_code_change_examples/horovod_mnist.py

```diff
@@ -0,0 +1,198 @@
+                test_loss, 100.0 * test_accuracy
+            )
+        )
+
+
+if __name__ == "__main__":
+    # Training settings
+    parser = argparse.ArgumentParser(description="PyTorch MNIST Example")
+    parser.add_argument(
+        "--epochs", type=int, default=1, metavar="N", help="number of epochs to train (default: 1)"
+    )
+    parser.add_argument("--use_only_cpu", type=str2bool, default=False)
+    parser.add_argument(
+        "--log-interval",
+        type=int,
+        default=10,
+        metavar="N",
+        help="how many batches to wait before logging training status",
+    )
+    args = parser.parse_args()
+    args.cuda = not args.use_only_cpu and torch.cuda.is_available()
+
+    # constants
+    batch_size = 64
+    lr = 0.01
+    momentum = 0.5
+    seed = 42
+    test_batch_size = 1000
+
+    # Horovod: initialize library.
+    hvd.init()
+    torch.manual_seed(seed)
+
+    if args.cuda:
+        # Horovod: pin GPU to local rank.
+        torch.cuda.set_device(hvd.local_rank())
+        torch.cuda.manual_seed(seed)
+
+    kwargs = {"num_workers": 1, "pin_memory": True} if args.cuda else {}
+    train_dataset = datasets.MNIST(
+        "data-%d" % hvd.rank(),
+        train=True,
+        download=True,
+        transform=transforms.Compose(
+            [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
+        ),
+    )
+    # Horovod: use DistributedSampler to partition the training data.
+    train_sampler = torch.utils.data.distributed.DistributedSampler(
+        train_dataset, num_replicas=hvd.size(), rank=hvd.rank()
+    )
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset, batch_size=batch_size, sampler=train_sampler, **kwargs
+    )
+
+    test_dataset = datasets.MNIST(
+        "data-%d" % hvd.rank(),
+        train=False,
+        transform=transforms.Compose(
+            [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
+        ),
+    )
+    # Horovod: use DistributedSampler to partition the test data.
+    test_sampler = torch.utils.data.distributed.DistributedSampler(
+        test_dataset, num_replicas=hvd.size(), rank=hvd.rank()
+    )
+    test_loader = torch.utils.data.DataLoader(
+        test_dataset, batch_size=test_batch_size, sampler=test_sampler, **kwargs
+    )
+
+    model = Net()
+
+    if args.cuda:
+        # Move model to GPU.
+        model.cuda()
+
+    # Horovod: scale learning rate by the number of GPUs.
+    optimizer = optim.SGD(model.parameters(), lr=lr * hvd.size(), momentum=momentum)
+
+    # Horovod: broadcast parameters & optimizer state.
+    hvd.broadcast_parameters(model.state_dict(), root_rank=0)
+    hvd.broadcast_optimizer_state(optimizer, root_rank=0)
+
+    # Horovod: (optional) compression algorithm.
+    compression = hvd.Compression.none
+
+    # Horovod: wrap optimizer with DistributedOptimizer.
+    optimizer = hvd.DistributedOptimizer(
```

Comment:
optimizer is not passed to train function ? 
train_function uses this optimizer in L53 . 
Wondering how is this working? Is this global static object?

Reply:
Interesting question.

I double checked with the [example](https://github.com/horovod/horovod/blob/master/examples/mxnet_mnist.py) and they seem to be doing the same thing.

I also defined an optimizer outside of the train function with a vanilla pytorch example and it failed with `NameError: name 'optimizer' is not defined` as expected.

I think in this case, hvd is taking care of the distribution and availability of the optimizer, since we have run `hvd.init()` before the definition of the optimizer and we also wrap the optimizer with `hvd.DistributedOptimizer`

Reply:
Then all examples are wrong, probably?
"I think in this case, hvd is taking care of the distribution and availability of the optimizer, since we have run hvd.init() before the definition of the optimizer and we also wrap the optimizer with hvd.DistributedOptimizer"

I didn't understand this. 
There is local variable optimizer, please figure out the right way to do it. Or say optimizer is global ?
Who is creating that global object ? 
Does it still work if you change optimizer to optimizer1 in all your source code?

Reply:
Looks like its a feature of python. 

Links: 
[DOCS](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces)
[S/O answer](https://stackoverflow.com/questions/4775579/main-and-scoping-in-python).


I tested this snippet in Python 3


```
def test_random_var():
    assert __name__ == "__main__"
    assert random_var == 5008
    assert 'random_var' in globals()
    print("Pass")

if __name__ ==  "__main__":
    random_var = 5008
    test_random_var()
```

It succeeds when we run `python test.py` and fails for `pytest test.py`.
Because `__name__ == "test"` in the latter.